src/qtism/common/datatypes/files/FileHash.php

```diff
@@ -0,0 +1,281 @@
+    {
+        $this->setId($id);
+        $this->setMimeType($mimeType);
+        $this->setFilename($filename);
+        $this->setHash($hash);
+    }
+
+    /**
+     * Create a new FileHash object from an array of properties.
+     *
+     * @param array $properties
+     * @return FileHash
+     */
+    public static function createFromArray(array $properties): self
+    {
+        return new self(
+            $properties['id'],
+            $properties['mime'],
+            $properties['name'],
+            $properties['data']
+        );
+    }
+
+    public function jsonSerialize(): array
+    {
+        return [
+            'id' => $this->id,
+            'mime' => $this->mimeType,
+            'name' => $this->filename,
+            'data' => $this->hash,
+        ];
+    }
+
+    /**
+     * Set the id of the file.
+     *
+     * @param string $id
+     */
+    protected function setId($id)
+    {
+        $this->id = $id;
+    }
+
+    /**
+     * Get the id of the file.
+     *
+     * @return string
+     */
+    public function getId()
+    {
+        return $this->id;
+    }
+
+    /**
+     * Set the mime-type of the file.
+     *
+     * @param string $mimeType
+     */
+    protected function setMimeType($mimeType)
+    {
+        $this->mimeType = $mimeType;
+    }
+
+    /**
+     * Get the mime-type of the file.
+     *
+     * @return string
+     */
+    public function getMimeType()
+    {
+        return $this->mimeType;
+    }
+
+    /**
+     * Get the name of the file.
+     *
+     * @return string
+     */
+    public function getFilename()
+    {
+        return $this->filename;
+    }
+
+    /**
+     * Set the name of the file.
+     *
+     * @param string $filename
+     */
+    protected function setFilename($filename)
+    {
+        $this->filename = $filename;
+    }
+
+    /**
+     * Get the sequence of bytes composing the hash of the file.
+     */
+    public function getData()
+    {
+        return $this->getHash();
+    }
+
+    /**
+     * Returns nothing because the content of the file is stored externally
+     * and thus not accessible in here.
+     */
+    public function getStream()
+    {
+    }
+
+    /**
+     * Get the hash of the file.
+     *
+     * @return string
+     */
+    public function getHash()
+    {
+        return $this->hash;
+    }
+
+    /**
+     * Set the hash of the file.
+     *
+     * @param string $hash
+     */
+    protected function setHash($hash)
+    {
+        $this->hash = $hash;
+    }
+
+    /**
+     * Get the cardinality of the File value.
+     *
+     * @return int A value from the Cardinality enumeration.
+     */
+    public function getCardinality()
+    {
+        return Cardinality::SINGLE;
+    }
+
+    /**
+     * Get the baseType of the File value.
+     *
+     * @return int A value from the BaseType enumeration.
+     */
+    public function getBaseType()
+    {
+        return BaseType::FILE;
```

Comment:
Shouldn't it be file hash instead of file?

Reply:
It is still a QTi datatype File. The FileHash is not in the standard, it's just an alternative implementation, and since we have some existing tests with uploaded files we can not just change the way we store them, that would make all the old test not being retrievable.

Reply:
okay thanks for the explanation :)